### DISCUSSION TITLE
selected rows

### DISCUSSION THREAD

**Discussion by photojo** [filamentphp/filament#5664] [Help]
How can I get just the selected rows in table? 

What is wrong with 
`dd($this->selectedTableRecords);`
?!

**Comment by danharrin** [comment]
Selected rows are handled in JS and only sent to PHP when a bulk action is run, otherwise the whole page slows down with lots of rows selected

**Comment by photojo** [comment]
But how can I pass only the selected rows to my Exportservice
`return (new Exportservice())->download('Export.xlsx', Excel::XLSX, ['Content-Type' => 'text/csv']);`

**Reply by danharrin** [reply]
`$records` inside the `BulkAction::make()->action(function ($records) {...})`

**Comment by photojo** [comment]
Thank you so much for helping me, Dan. The bulk-action is not the problem. 
I can not pass **some single selected rows** to my Exportservice. 

`ExportAction::make() ->exports([ExcelExport::make('model') ...` works fine, but the columns are not in the order, which I want, although I use 
`->withColumns([Column::make() ...`


So, I made my own Export 
`Action::make('export excel xlsx')
                ->label('Export Excel')
                ->button()
                ->action(function () {return (new Exportservice())->download('Export.xlsx', Excel::XLSX, ['Content-Type' => 'text/csv']);
});

`

In my Exportservice I have

`public function collection()
    {
        return Users::all();
    }`


But I only get all rows OR filtered rows. But I have to select some rows with left selectionfield.

I'm so sorry for asking maybe simple questions/solutions. And for my english.

**Reply by ArtMin96** [reply]
Dan has already demonstrated this: `BulkAction::make()->action(function ($records) {...})`. The `$records` written in the action callback are the selected rows.
```php
YourBulkAction::make()
    ->action(function ($records) {
        // Do whatever you want with selected records.
        dd($records);
    })
```

**Reply by photojo** [reply]
unfortunately, with 
```
protected function getTableHeaderActions()
    {
        return [
            Action::make('export')
                ->button()
                ->action(function (Collection $records) {
                    dd($records);
                }),
```
I get no results, if I select one or more then one or all records?!

**Reply by ArtMin96** [reply]
Are you using admin panel or table package?

**Reply by ArtMin96** [reply]
https://filamentphp.com/docs/2.x/tables/actions#bulk-actions

**Reply by photojo** [reply]
I use table package.

**Reply by photojo** [reply]
Oh my goodness. Thank you very much! 
I think, that was a misunderstanding, because I want to have the Export-Action-Button as a separate button in table-header ( protected function getTableHeaderActions() ...)

I think, I have to go with the Bulkaction, haven't I?!

**Reply by ArtMin96** [reply]
```php
protected function getTableBulkActions(): array
{
    return [
        BulkAction::make('action_name')
            ->action(fn (Collection $records) => dd($records))
    ];
}
```

**Reply by photojo** [reply]
Ok! Thank you very much again! I do it with getTableBulkActions() and remove ist from      getTableHeaderActions().

Bye!

**Reply by danharrin** [reply]
In v3 you can add bulk actions to the header

**Reply by photojo** [reply]
Ah, ok! Thank's a lot, Dan!

**Comment by photojo** [comment]
Is it possible to call the action automatically every time a row is selected?
